stream.go

```diff
@@ -48,27 +61,24 @@ type stream struct {
 	writeOffset    protocol.ByteCount
 	readOffset     protocol.ByteCount
 
-	// Once set, the errors must not be changed!
-	err error
+	closeForShutdownErr error
+	cancelWriteErr      error
+	cancelReadErr       error
+	resetRemotelyErr    StreamError
 
-	// closedForShutdown is set when Cancel() is called
-	closedForShutdown utils.AtomicBool
-	// finishedReading is set once we read a frame with a FinBit
-	finishedReading utils.AtomicBool
-	// finisedWriting is set once Close() is called
-	finishedWriting utils.AtomicBool
-	// resetLocally is set if Reset() is called
-	resetLocally utils.AtomicBool
-	// resetRemotely is set if HandleRstStreamFrame() is called
-	resetRemotely utils.AtomicBool
+	closedForShutdown bool // set when CloseForShutdown() is called
+	finRead           bool // set once we read a frame with a FinBit
+	finishedWriting   bool // set once Close() is called
+	canceledWrite     bool // set when CancelWrite() is called
+	canceledRead      bool // set when CancelRead() is called
+	finSent           bool // set when a STREAM_FRAME with FIN bit has b
+	resetRemotely     bool // set when HandleRstStreamFrame() is called
```

Comment:
This is non-blocking for this PR, but I'd really like to get rid of all these bools and replace them with a state machine enum and a single var in the stream. Let's discuss this during one of the next weeklies.